orttraining/orttraining/core/agent/training_agent.cc

```diff
@@ -2,135 +2,80 @@
+      break;
     }
-    ORT_ENFORCE(thread.joinable());
-    thread.join();
-    onnxruntime::contrib::OrtTasks::GetInstance().RemoveTask(run_id);
-    return forward_status;
+    break_point += 1;
   }
 
-  return Status::OK();
+  fw_program_counter_end_ = break_point - 1;
+  bw_program_counter_end_ = end_point;
 }
 
-common::Status TrainingAgent::RunBackward(int64_t run_id, const std::vector<OrtValue>& backward_output_grads) {
-  LOGS(*inference_session_.GetLogger(), VERBOSE) << "Running TrainingAgent::Backward() with run_id " << run_id;
-
-  // resume background thread
-  onnxruntime::contrib::OrtTasks::GetInstance().SetBackwardInputs(run_id, backward_output_grads, false);
-
-  Status bg_thread_status = onnxruntime::contrib::OrtTasks::GetInstance().WaitForStatus(run_id);
-
-  std::thread bg_thread;
-  {
-    std::lock_guard<std::mutex> lock(bg_threads_mutex_);
-    std::swap(bg_thread, bg_threads_[run_id]);
-    bg_threads_.erase(run_id);
-  }
-
-  // wait for bg_thread to complete
-  ORT_ENFORCE(bg_thread.joinable());
-  bg_thread.join();
-  onnxruntime::contrib::OrtTasks::GetInstance().RemoveTask(run_id);
+TrainingAgent::~TrainingAgent(){};
 
-  return bg_thread_status;
+void TrainingAgent::RunForward(const std::vector<OrtValue>& feeds, std::vector<OrtValue>& fetches, PartialGraphExecutionState& state) {
+  state.SetProgramCounterStart(0);
+  state.SetProgramCounterEnd(fw_program_counter_end_);
+  RunCore(feeds, fetches, state, *fw_feeds_fetches_manager_);
 }
 
-void TrainingAgent::CancelPendingBackwardRun(int64_t run_id) {
-  LOGS(*inference_session_.GetLogger(), INFO) << "Canceling background task with run_id " << run_id;
+void TrainingAgent::RunBackward(const std::vector<OrtValue>& feeds, std::vector<OrtValue>& fetches, PartialGraphExecutionState& state) {
+  state.SetProgramCounterStart(fw_program_counter_end_ + 1);
+  state.SetProgramCounterEnd(bw_program_counter_end_);
+  RunCore(feeds, fetches, state, *bw_feeds_fetches_manager_);
+}
 
-  // resume background thread with terminate = true
-  onnxruntime::contrib::OrtTasks::GetInstance().SetBackwardInputs(run_id, {}, true);
+void TrainingAgent::RunCore(const std::vector<OrtValue>& feeds, std::vector<OrtValue>& fetches, PartialGraphExecutionState& state, FeedsFetchesManager& feeds_fetches_manager) {
+  auto fetches_size = feeds_fetches_manager.GetFeedsFetchesInfo().output_names.size();
+  fetches.resize(fetches_size);
+  for (size_t index = 0; index < fetches_size; index += 1) {
+    fetches[index] = {};
+  }
 
-  // wait for bg_thread to complete
-  std::thread bg_thread;
-  {
-    std::lock_guard<std::mutex> lock(bg_threads_mutex_);
-    std::swap(bg_thread, bg_threads_[run_id]);
-    bg_threads_.erase(run_id);
+  RunOptions run_options;
+  auto status = inference_session_.PartialRun(run_options, feeds, fetches, state, feeds_fetches_manager);
+  if (!status.IsOK()) {
+    throw std::runtime_error("Error in execution: " + status.ErrorMessage());
```

Comment:
I see, you throw here...

Reply:
I think it's not too expensive to follow the old pattern. 
TraningAgent should return the status, and throw can happen in the python side.